SIL: Force Clang-imported protocol conformances to get deserialized when used.

jckarter opened the pull request:
Code may end up indirectly using a witness table for a Clang-imported type by inlining code that used the conformance from another module, in which case we need to ensure we have a local definition at hand in the inlining module so we can have something to link against independently. This needs to be fixed from both sides:

- During serialization, serialize not only witness tables from the current module, but from Clang-imported modules too
- During deserialization, when the SILLinker walks a loaded module, ensure that all shared conformances get deserialized, including those from ApplyInsts and inherited/associated type protocol requirements.

Fixes rdar://problem/38687726.

jckarter committed:
SIL: Force Clang-imported protocol conformances to get deserialized when used.

Code may end up indirectly using a witness table for a Clang-imported type by inlining code that used the conformance from another module, in which case we need to ensure we have a local definition at hand in the inlining module so we can have something to link against independently. This needs to be fixed from both sides:

- During serialization, serialize not only witness tables from the current module, but from Clang-imported modules too
- During deserialization, when the SILLinker walks a loaded module, ensure that all shared conformances get deserialized, including those from ApplyInsts and inherited/associated type protocol requirements.

Fixes rdar://problem/38687726.

jckarter committed:
Serialization: Never serialize conformances from Clang modules.

We should always go through the Clang importer to reinstantiate them. Serializing them leads to the possibility of us ending up with multiple conformances with different identities that ought to be equivalent.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test compiler performance

jckarter commented:
@slavapestov I went back to the eager deserialization approach since trying to do it lazily in IRGen is too late due to pre-IRGen lowering passes on SIL. How does this look (assuming the performance impact is acceptable)?

slavapestov commented on test/IRGen/Inputs/deserialize-clang-importer-witness-tables/clang-importer-witness-tables/regex.swift:
```diff
@@ -0,0 +1 @@
+pbpaste > test/IRGen/Inputs/deserialize/clang-importer-witness-tables/regex.swift
```
Did you mean to heck this in?

jckarter committed:
SIL deserializer: Allow multiple shared definitions of a witness table.

By design, things with shared linkage are allowed to have definitions in different TUs, and we're supposed to pick one when linking or merging modules.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test compiler performance

jckarter commented:
@swift-ci Please test source compatibility

jckarter commented:
@swift-ci Please test compiler performance
